File: server/engines/apple_vision.py
import json
import time
import subprocess
from pathlib import Path

BINARY_PATH = Path(__file__).resolve().parent.parent.parent / "tools" / "apple_vision_ocr"


def run_apple_vision_ocr(image_path: str) -> dict:
    start_time = time.perf_counter()
    if not BINARY_PATH.exists():
        raise FileNotFoundError(f"Apple Vision OCR binary not found at: {BINARY_PATH}")

    result = subprocess.run(
        [str(BINARY_PATH), str(image_path)],
        capture_output=True,
        text=True,
        check=False
    )
    fallback_latency = (time.perf_counter() - start_time) * 1000

    if result.returncode != 0 and not result.stdout.strip():
        raise RuntimeError(f"Apple Vision OCR execution error: {result.stderr or 'Process failed'}")

    try:
        parsed = json.loads(result.stdout)
        if parsed.get("error"):
            raise RuntimeError(parsed["error"])

        latency_ms = round(float(parsed.get("latency_ms", fallback_latency)), 1)
        return {
            "text": parsed.get("text", ""),
            "lines": parsed.get("lines", []),
            "latencyMs": latency_ms,
            "engine": "apple_vision",
            "engineName": "Apple Vision OCR (macOS Native)"
        }
    except json.JSONDecodeError as e:
        raise RuntimeError(f"Failed to parse Apple Vision output: {e}. Raw: {result.stdout}")

File: server/engines/appleVision.js
import { execFile } from 'child_process';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const BINARY_PATH = path.resolve(__dirname, '../../tools/apple_vision_ocr');

export async function runAppleVisionOCR(imagePath) {
  return new Promise((resolve, reject) => {
    const startTime = performance.now();
    execFile(BINARY_PATH, [imagePath], { maxBuffer: 10 * 1024 * 1024 }, (error, stdout, stderr) => {
      const fallbackLatency = performance.now() - startTime;
      if (error) {
        return reject(new Error(`Apple Vision OCR execution error: ${stderr || error.message}`));
      }

      try {
        const parsed = JSON.parse(stdout);
        if (parsed.error) {
          return reject(new Error(parsed.error));
        }
        resolve({
          text: parsed.text || '',
          lines: parsed.lines || [],
          latencyMs: Math.round((parsed.latency_ms || fallbackLatency) * 10) / 10,
          engine: 'apple_vision',
          engineName: 'Apple Vision OCR (macOS Native)'
        });
      } catch (err) {
        reject(new Error(`Failed to parse Apple Vision output: ${err.message}. Raw: ${stdout}`));
      }
    });
  });
}
